src/estimagic/optimization/tranquilo/tranquilo.py

```diff
@@ -260,97 +309,201 @@ def _tranquilo(
-
+        # ==========================================================================
+        # sample points if necessary and do simple iteration
+        # ==========================================================================
         new_xs = sample_points(
             trustregion=state.trustregion,
-            n_points=n_to_sample,
-            existing_xs=filtered_xs,
+            n_points=max(0, target_sample_size - len(model_xs)),
+            existing_xs=model_xs,
             rng=sampling_rng,
         )
 
-        # ==============================================================================
-        # criterion evaluations
-        # ==============================================================================
-
-        _, _, new_indices = wrapped_criterion(new_xs)
-        model_indices = np.hstack([filtered_indices, new_indices])
-        model_xs = history.get_xs(model_indices)
-        model_fvecs = history.get_fvecs(model_indices)
+        new_indices = history.add_xs(new_xs)
 
-        # ==============================================================================
-        # build surrogate and optimize it
-        # ==============================================================================
+        eval_info = {i: n_evals_per_point for i in new_indices}
 
-        weights = calculate_weights(model_xs, trustregion=state.trustregion)
+        evaluate_criterion(eval_info)
 
-        centered_xs = (model_xs - state.trustregion.center) / state.trustregion.radius
+        model_indices = _concatenate_indices(model_indices, new_indices)
 
-        clipped_fvecs = clip_infinite_values(model_fvecs)
+        model_xs = history.get_xs(model_indices)
+        model_data = history.get_model_data(
+            x_indices=model_indices,
+            average=True,
+        )
 
-        vector_model = fit_model(centered_xs, clipped_fvecs, weights=weights)
+        vector_model = fit_model(
+            *model_data,
+            region=state.trustregion,
+            old_model=state.vector_model,
+            weights=None,
+        )
 
         scalar_model = aggregate_vector_model(
             vector_model=vector_model,
         )
 
         sub_sol = solve_subproblem(model=scalar_model, trustregion=state.trustregion)
 
+        _relative_step_length = (
+            np.linalg.norm(sub_sol.x - state.x) / state.trustregion.radius
+        )
+
+        # ==========================================================================
+        # If we have enough points, drop points until the relative step length
+        # becomes large enough
+        # ==========================================================================
+
+        if len(model_xs) > target_sample_size:
+            small_step = (
+                _relative_step_length < stagnation_options.min_relative_step_keep
+            )
+
+            while small_step and len(model_xs) > target_sample_size:
```

Comment:
`small_step` needs to be updated inside the while loop

Reply:
Done.